gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java

```diff
@@ -0,0 +1,191 @@
+
+  // Do cleanup the path after every test.
+  private static final String GCS_WRITE_TMP_DIR = "/tmp/gcs-write-dir";
+  private static final String GCS_WRITE_TMP_DIR_1 = "/tmp/gcs-write-dir-1";
+  /**
+   * Deafult value is picked from java-storage API.
+   *
+   * @see BlobWriteSessionConfigs#bufferToTempDirThenUpload()
+   */
+  private static final String DEFAULT_TEMP_SUB_DIR = "google-cloud-storage";
+
+  private static GoogleCloudStorage helperGcs;
+
+  @Rule
+  public TestName name =
+      new TestName() {
+        // With parametrization method name will get [index] appended in their name.
+        @Override
+        public String getMethodName() {
+          return super.getMethodName().replaceAll("[\\[,\\]]", "");
+        }
+      };
+
+  @BeforeClass
+  public static void before() throws IOException {
+
+    helperGcs = GoogleCloudStorageTestHelper.createGcsClientImpl();
+    helperGcs.createBucket(TEST_BUCKET);
+  }
+
+  @AfterClass
+  public static void after() throws IOException {
+    try {
+      BUCKET_HELPER.cleanup(helperGcs);
+    } finally {
+      helperGcs.close();
+    }
+  }
+
+  @Before
+  public void setUp() {
+    System.setProperty("java.io.tmpdir", GCS_WRITE_TMP_DIR);
+  }
+
+  @After
+  public void cleanUp() {
+    ImmutableSet<String> tempDirs = ImmutableSet.of(GCS_WRITE_TMP_DIR_1, GCS_WRITE_TMP_DIR);
+    Iterator<String> iterator = tempDirs.stream().iterator();
+    while (iterator.hasNext()) {
+      String filePath = iterator.next();
+      for (File file : new File(filePath).listFiles()) {
+        file.delete();
+      }
+    }
+  }
+
+  @Test
+  public void writeToDefaultPathThenUploadEnabled() throws IOException {
+    GoogleCloudStorageOptions storageOptions =
+        GoogleCloudStorageTestHelper.getStandardOptionBuilder()
+            .setWriteChannelOptions(
+                AsyncWriteChannelOptions.builder().setDiskThenUploadEnabled(true).build())
+            .build();
+
+    GoogleCloudStorage gcs = getGCSImpl(storageOptions);
+    StorageResourceId resourceId = new StorageResourceId(TEST_BUCKET, name.getMethodName());
+
+    byte[] bytesToWrite = new byte[1024];
+    GoogleCloudStorageTestHelper.fillBytes(bytesToWrite);
+    Path temporaryPath = Paths.get(GCS_WRITE_TMP_DIR, DEFAULT_TEMP_SUB_DIR);
+    // validate that there were no files
+    verifyTemporaryFileCount(ImmutableSet.of(temporaryPath), 0);
+
+    WritableByteChannel writeChannel = gcs.create(resourceId);
+    writeChannel.write(ByteBuffer.wrap(bytesToWrite));
+    // temporary files created in disk.
+    verifyTemporaryFileCount(ImmutableSet.of(temporaryPath), 1);
+
+    writeChannel.close();
+    // temporary files will be deleted from disk once upload is finished.
+    verifyTemporaryFileCount(ImmutableSet.of(temporaryPath), 0);
+
+    gcs.close();
+  }
+
+  @Test
+  public void writeToPathThenUploadEnabled() throws IOException {
+    ImmutableSet<String> tempDirs = ImmutableSet.of(GCS_WRITE_TMP_DIR_1, GCS_WRITE_TMP_DIR);
```

Comment:
consider doing code refactoring and getting rid of code duplication.

Reply:
done